lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -1664,14 +1664,44 @@ Value *PatchInOutImportExport::patchFsGenericInputImport(Type *inputTy, unsigned
   const unsigned locCount = inputTy->getPrimitiveSizeInBits() / 8 > SizeOfVec4 ? 2 : 1;
   while (interpInfo.size() <= location + locCount - 1)
     interpInfo.push_back(InvalidFsInterpInfo);
+
+  // Check last vertex processing stage to see if there is an output matching this input.
+  bool noVsMatch = true;
+  auto lastVertexStage = m_pipelineState->getLastVertexProcessingStage();
+  lastVertexStage = lastVertexStage == ShaderStageCopyShader ? ShaderStageGeometry : lastVertexStage;
+  if (lastVertexStage != ShaderStageInvalid) {
+    const auto &lastVertexInOutUsage = m_pipelineState->getShaderResourceUsage(lastVertexStage)->inOutUsage;
+    const auto streamId = lastVertexStage == ShaderStageGeometry ? lastVertexInOutUsage.gs.rasterStream : 0;
+
+    // We first check generic outputs of last vertex processing stage
+    for (auto &locMap : lastVertexInOutUsage.outputLocInfoMap) {
+      if (locMap.second.getLocation() == location && locMap.second.getStreamId() == streamId) {
+        noVsMatch = false;
+        break;
+      }
+    }
+
+    // If we don't find a matching output, we then check built-in outputs of last vertex processing stage. Such
+    // built-ins are exported as generic outputs from the last vertex processing stage like gl_PrimitiveID,
+    // gl_Layer, gl_ViewportIndex, and gl_ViewIndex.
+    if (noVsMatch) {
+      for (auto &locMap : lastVertexInOutUsage.builtInOutputLocMap) {
+        if (locMap.second == location) {
+          noVsMatch = false;
```

Comment:
What will happen in an unlinked pixel shader?  Since there are no other shader stages, noVsMatch will always be true, even if there is a match after linking the pipeline.  Will this cause a problem?

Reply:
You are right. This could cause problems when we link those ELFs finally. The in/out matching info is missing in ELF. A way to detect this is to decode the targets of exp instructions in HW VS. Do you have any suggestions?

Reply:
I was thinking about this yesterday.  Unlinked shader compilation assumes that the interfaces match exactly.  If they do not then, other really bad things could happen.  As an example suppose we have a vertex shader with outputs in SPIR-V location 10, 20, and 30.  Those will be mapped to HW locations 0, 1, and 2.  Further suppose the PS has only 2 inputs at spir-v locations 10 and 30.  Then llpc will map these to HW locations 0 and 1.  Notice there will now be a mismatch.  The output at spir-v location 20 will be written in the location the PS expects the input at spir-v location 30.

For this reason, I think we should disable relocatable shader compilation when there is a mismatch, and fall back to a full pipeline compilation.

If this become a big enough issue, we should probably add relocations or something for the HW locations and replace those value while linking.  I don't expect you to do that.

Reply:
Thank you. It's clear to me now. This issue is found by HW simulator as an assertion. Real HW doesn't report any problem. It seems to be a small issue. I will further confirm this. If we can relax the assertion, I will drop this change.